crates/unleash-edge/src/edge_builder.rs

```diff
@@ -210,8 +222,213 @@ pub async fn build_edge(
+    );
+
+    let app_state = AppState::builder()
+        .with_token_cache(token_cache.clone())
+        .with_features_cache(features_cache.clone())
+        .with_engine_cache(engine_cache.clone())
+        .with_token_validator(Arc::new(Some(token_validator.as_ref().clone())))
+        .with_feature_refresher(Arc::new(Some(feature_refresher.as_ref().clone())))
+        .with_metrics_cache(metrics_cache.clone())
+        .with_persistence(persistence)
+        .with_deny_list(args.http.deny_list.unwrap_or_default())
+        .with_allow_list(args.http.allow_list.unwrap_or_default())
+        .with_instance_sending(instance_data_sender)
+        .with_edge_instance_data(edge_instance_data)
+        .build();
+
+    Ok((app_state, background_tasks))
+}
+
+fn create_edge_mode_background_tasks(
+    edge: EdgeArgs,
+    client_meta_information: ClientMetaInformation,
+    feature_refresher: Arc<FeatureRefresher>,
+    persistence: Option<Arc<dyn EdgePersistence>>,
+    token_cache: Arc<TokenCache>,
+    feature_cache: Arc<FeatureCache>,
+    validator: Arc<TokenValidator>,
+    http_client: reqwest::Client,
+    registry: prometheus::Registry,
+    app_name: String,
+    instance_data_sender: Arc<InstanceDataSending>,
+    edge_instance_data: Arc<EdgeInstanceData>,
+    instances_observed_for_app_context: Arc<RwLock<Vec<EdgeInstanceData>>>,
+    metrics_cache_clone: Arc<MetricsCache>,
+    unleash_client: Arc<UnleashClient>,
+) -> Vec<Pin<Box<dyn Future<Output = ()> + Send>>> {
+    let mut tasks: Vec<Pin<Box<dyn Future<Output = ()> + Send>>> = vec![];
+    tasks.push(create_fetch_task(
+        &edge,
+        client_meta_information,
+        feature_refresher.clone(),
+    ));
+
+    tasks.push(create_send_metrics_task(
+        metrics_cache_clone.clone(),
+        unleash_client,
+        token_cache.clone(),
+        edge.metrics_interval_seconds.try_into().unwrap(),
+    ));
+
+    tasks.push(create_revalidation_task(
+        &validator,
+        edge.token_revalidation_interval_seconds,
+    ));
+
+    tasks.push(create_revalidation_of_startup_tokens_task(
+        &validator,
+        edge.tokens.clone(),
+        feature_refresher.clone(),
+    ));
+
+    tasks.push(create_send_instance_data_task(
+        instance_data_sender.clone(),
+        edge_instance_data.clone(),
+        instances_observed_for_app_context.clone(),
+    ));
+
+    if let Some(url) = edge.prometheus_remote_write_url {
+        tasks.push(create_prometheus_write_task(
+            http_client,
+            registry,
+            url,
+            edge.prometheus_push_interval,
+            app_name,
+        ));
+    }
+
+    if let Some(persistence) = persistence {
+        tasks.push(create_persist_data_task(
+            persistence.clone(),
+            token_cache.clone(),
+            feature_cache.clone(),
+        ));
+    } else {
+        info!("No persistence configured, skipping persistence");
+    }
+
+    tasks
+}
+
+fn create_fetch_task(
```

Comment:
Broke this out into it's own task. This was split over 3 places, with a busy loop for fetching if you were actually streaming. Now we make one decision on how to fetch features and all the options are mutually exclusive. You can no longer theoretically be both polling and streaming